### PR TITLE
Mark document as printed

### LINKED ISSUE #198: Create order filter for printing actions
Currently we don't have a way to know when a document is printed, but would be nice if we could do that and also add a filter in the orders list to get printed or non-printed orders for each type of document.

Forum ticket: https://wordpress.org/support/topic/new-filter-in-orders-page/?view=all

### PR DESCRIPTION
**Pull request by alexmigf**
closes #198 

This PR adds 2 new document settings, for now only for **Invoice**:

Captura de ecrã de 2023-02-27 14-59-01

Captura de ecrã de 2023-02-27 14-58-46

Selecting a **Mark printed** option beside `no` or `manually`, it will mark the document printed automatically in the selected event:

Captura de ecrã de 2023-02-27 15-01-35

Captura de ecrã de 2023-02-27 15-01-47

If **Unmark printed** is `enabled` a link with action is added:

Captura de ecrã de 2023-02-27 15-10-26

If **Mark printed** is set to `manually`, a **Mark printed** link will be display underneath the Invoice button:

Captura de ecrã de 2023-02-27 15-12-35

If **Log to order notes** is **enable** under the Status tab, the mark/unmark printed events are registered in the order notes:

Captura de ecrã de 2023-02-27 15-14-19

### CODE REVIEW

**Comment by dwalkerpriv** [comment]
So far I've selected the "manually" printed option only.
the action link to mark as printed appears
click the link to mark as printed redirects me to this page with a 0 on it:

Screenshot 2023-02-27 110456

Going back to the previous page, the print status is updated:
Screenshot 2023-02-27 110745

There are some typos in the dropdown options with an extra "on" in the option name

Screenshot 2023-02-27 110144

**Comment by alexmigf** [comment]
@dwalkerpriv I'm not having any of those issues. Do you have the branch updated? Perhaps you have the extension running twice with a different name?

**Comment by dwalkerpriv** [comment]
In the latest pull, the typos are gone. The options show normal now:

Screenshot 2023-02-28 090244

Clicking the 'mark printed' link still brings me to a page with a 0 on it. How would I check if I'm running another instance of the extension?

**Testing the "Single order action":**
- Print icon doesn't up when the invoice action button is clicked on the Orders page.
- Switching from one option to the other removes the printed icon if the order was marked printed already. Is this intended?
- Switching back to the previous option, shows the print icon/status.
- After switching to a previous option where an invoice was marked printed, clicking the 'unmark print' link redirects me to the page with a 0 also

I haven't found anything in the error logs related to this as yet either

**Comment by alexmigf** [comment]
@dwalkerpriv did you cleared the browser cache?

**Comment by dwalkerpriv** [comment]
Yeah, I did. I haven't seen a difference yet

**Comment by alexmigf** [comment]
@dwalkerpriv do you spot anything in the browser JS console?

**Comment by dwalkerpriv** [comment]
In manual mode, when I click 'unmark printed' I get this in the console before it redirects the page with 0

Screenshot 2023-02-28 093622

The same error occurs when I click 'mark printed', along with the redirect to the page

**Comment by alexmigf** [comment]
Could you enable `XHR` and see if the request catches any error?

**Comment by dwalkerpriv** [comment]
This is before I clicked unmark:

Screenshot 2023-02-28 094233

After:

Screenshot 2023-02-28 094301

The same happens when I click 'mark printed' again

**Comment by alexmigf** [comment]
I'm not seeing there any of the 2 requests, `mark` or `unmark`.

**Comment by dwalkerpriv** [comment]
In a private window I did a test and got this in the response:

```GEThttps://wposetup.local/wp-admin/admin-ajax.php?action=printed_wpo_wcpdf&event=mark&document_type=invoice&order_id=790&trigger=manually&security=0e25d4c089
[HTTP/2 200 OK 1617ms]

	
GET
	https://wposetup.local/wp-admin/admin-ajax.php?action=printed_wpo_wcpdf&event=mark&document_type=invoice&order_id=790&trigger=manually&security=0e25d4c089
Status
200
OK
VersionHTTP/2
Transferred556 B (1 B size)
Referrer Policystrict-origin-when-cross-origin
Request PriorityHighest

This page is in Quirks Mode. Page layout may be impacted. For Standards Mode use “<!DOCTYPE html>”. admin-ajax.php
POSThttps://wposetup.local/wp-admin/admin-ajax.php
[HTTP/2 200 OK 1782ms]

GEThttps://wposetup.local/favicon.ico
[HTTP/2 302 Found 0ms]

GEThttps://wposetup.local/wp-includes/images/w-logo-blue-white-bg.png
[HTTP/2 200 OK 0ms]```

**Comment by alexmigf** [comment]
@dwalkerpriv could you test in a new setup?

**Comment by dwalkerpriv** [comment]
Sure

File: includes/documents/class-wcpdf-invoice.php [hunk 1]
```diff
@@ -411,6 +413,37 @@ public function init_settings() {
...
+			array(
+				'type'			=> 'setting',
+				'id'			=> 'mark_printed',
+				'title'			=> __( 'Mark printed', 'woocommerce-pdf-invoices-packing-slips' ),
+				'callback'		=> 'select',
+				'section'		=> 'invoice',
+				'args'			=> array(
+					'option_name' => $option_name,
+					'id'          => 'mark_printed',
+					'options'     => array_merge(
```
Comment (YordanSoares) [inline includes/documents/class-wcpdf-invoice.php:425]: I would add a description to this setting too, e.g. _"If enabled, allows you to mark the document as printed, manually (in the order page) or automatically (based on the document creation context you have selected)."_

@dwalkerpriv @dpeyou, what do you think?

**Review by YordanSoares — CHANGES_REQUESTED** [review]
@alexmigf I ran several tests from yesterday, and I have found a couple of issues:

1. The **On email attachment** option seems to be not working. I tried with several email notifications and the printed mark were not added.
2. The **On order document data** option stores the printed mark if you create the PDF invoices within the **PDF document data** panel (setting the invoice number and date), but it is not added for existing PDF invoices if you edit its number or date.

File: includes/documents/class-wcpdf-invoice.php [hunk 2]
```diff
@@ -411,6 +413,37 @@ public function init_settings() {
...
+				'type'			=> 'setting',
+				'id'			=> 'unmark_printed',
+				'title'			=> __( 'Unmark printed', 'woocommerce-pdf-invoices-packing-slips' ),
+				'callback'		=> 'checkbox',
+				'section'		=> 'invoice',
+				'args'			=> array(
+					'option_name'	=> $option_name,
+					'id'			=> 'unmark_printed',
+					/* translators: zero number */
+					'description'	=> __( 'Allow document to be unmark as printed.', 'woocommerce-pdf-invoices-packing-slips' ),
```
[the remark this answers is not in the document]
Comment (YordanSoares) [reply includes/documents/class-wcpdf-invoice.php]: I agree. I'll mark this conversation as resolved.

**Comment by alexmigf** [comment]
@YordanSoares I have fixed the issue with attachments and also your other comments, please confirm that everything is fine now.

File: includes/documents/class-wcpdf-invoice.php (hunk 1 above)
Comment (dwalkerpriv) [reply includes/documents/class-wcpdf-invoice.php:423]: @YordanSoares yes, this would be good and the description you provided here is well also 👍

**Review by YordanSoares — APPROVED** [review]
@alexmigf the printed mark is now displayed correctly when the document is created for the email attachment!

This is an exciting new improvement: Great job! 🎉

**Review by dwalkerpriv — APPROVED** [review]
Uploads working as with the prior tests. the destination folder is there as well

**Review by dwalkerpriv — CHANGES_REQUESTED** [review]
wrong PR approval

**Comment by dwalkerpriv** [comment]
@alexmigf on a new setup i'm still getting the redirect issue. perhaps its config in my Local (desktop app) config?

**Comment by alexmigf** [comment]
> @alexmigf I am getting the issue Yordan first encountered:
> 
> > The On email attachment option seems to be not working. I tried with several email notifications and the printed mark were not added.
> 
> I have done `git pull`, and **Umark as Printed** is set to true.
> 
> Also, **On Single Order Action**: is the intention that if you click the Invoice button, it should be marked as Printed? It isn't for me. I have verified that the Invoice does get attached.
> 
> I do not have errors in my WC logs.

Using the last commit you shouldn't have the issue. You can enable the log to order notes in the status page and see if it gets registered.

**Comment by alexmigf** [comment]
Made changes to support multiple mark printed actions:

Captura de ecrã de 2023-03-10 15-51-16

**Comment by alexmigf** [comment]
@dwalkerpriv @dpeyou @YordanSoares could you review again follow the last change mentioned [here](https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/483#issuecomment-1464007654)?

**Comment by dwalkerpriv** [comment]
@alexmigf so far, the markings work as expected. i still have the page redirect, though. i haven't noticed any other weirdness/errors as yet

**Comment by dwalkerpriv** [comment]
@alexmigf same on a new setup. i'm convinced it's something to do with my setup in Local:

Screenshot 2023-03-16 093624

I'm trying some new settings here now

**Comment by dwalkerpriv** [comment]
@alexmigf changing my server from apache to nginx fixed the issue. i can mark and unmark the pdfs as printed without the page redirect

**Comment by alexmigf** [comment]
@dwalkerpriv don't forget to leave your review please.

**Comment by dpeyou** [comment]
@alexmigf now it seems to only be working if the document has not been generated already.
So if the document has already been generated, it does not automatically mark the document.

**Comment by alexmigf** [comment]
@dwalkerpriv if the mark option is not manual it's correct, you can't mark a document on generation if it's already generated.

**Comment by Terminator-Barbapapa** [comment]
At the moment I can't mark a document as printed manually for already existing documents. Is that the desired behaviour?

I chose 'Manually' for the invoice and then went to an order to markt it as printed, but the link never showed. It took me a while to figure out I would first have to regenerate the invoice before the link showed up. I think that is confusing.

**Comment by alexmigf** [comment]
> At the moment I can't manually mark a document as printed for already existing documents. Is that the desired behaviour?

If you select to manual mark it should work, eg.

Screenshot from 2023-03-22 14-55-36

> I chose 'Manually' for the invoice and then went to an order to markt it as printed, but the link never showed. It took me a while to figure out I would first have to regenerate the invoice before the link showed up. I think that is confusing.

Not sure I agree with that, but maybe it's an edge case? What's the point on marking a document has printed when is not even generated yet? It's the same principal as the Credit Note button isn't it?

**Comment by alexmigf** [comment]
@Terminator-Barbapapa fixed

**Comment by alexmigf** [comment]
@Terminator-Barbapapa fixed here https://github.com/wpovernight/woocommerce-pdf-invoices-packing-slips/pull/483/commits/dace766370785bbed4c8ec8ffbd1c35dd24a2c30

File: includes/class-wcpdf-settings-debug.php [hunk 3]
```diff
@@ -509,7 +509,7 @@ public function init_settings() {
 				'args'     => array(
 					'option_name' => $option_name,
 					'id'          => 'log_to_order_notes',
-					'description' => __( 'Log PDF document creation to order notes.', 'woocommerce-pdf-invoices-packing-slips' ),
+					'description' => __( 'Log PDF document creation and mark/unmark printed to order notes.', 'woocommerce-pdf-invoices-packing-slips' ),
```
Comment (Terminator-Barbapapa) [inline includes/class-wcpdf-settings-debug.php:512]: Would add 'as':

`Log PDF document creation and mark/unmark as printed to order notes.`

**Review by Terminator-Barbapapa — APPROVED** [review]
Approved. Cool new feature! 😄 🎉

**Pull request merged**